Created SumOfProducts subclass of ControlValues

NoureldinYosri opened the pull request:
This class allows the creation of control values that can't be factored into simple products hence solving #4512

tanujkhattar commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,80 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
         return type(self)(self._internal_representation + other._internal_representation)
+
+
+@dataclass(frozen=True, eq=False)
+class SumOfProducts(AbstractControlValues):
+    """SumOfProducts represents control values in a form of a list or products.
+
+    SumOfProducts allows the creation of control values combinations that
+    can't be factored as the product of independent expressions (e.g. XOR)
+
+
+    Examples:
+        xor = SumOfProducts(((0, 1), (1, 0)))   # xor of two bits.
+        nand =  SumOfProducts(((0, 0), (0, 1), (1, 0))) # nand of two bits
+    """
```
```suggestion

@dataclass(frozen=True, eq=False)
class SumOfProducts(AbstractControlValues):
    """Represents control values as a union of n-bit tuples, each describing an allowed assignment.

    `SumOfProducts` representation describes the control values as a union of n-bit tuples, where
    each n-bit tuple represents an allowed assignment of bits for which the control should be
    activated. This expanded representation allows us to create control values combinations which
    cannot be factored as a `ProductOfSums` representation.

    For example:

    1) `(|00><00| + |11><11|) X + (|01><01| + |10><10|) I` represents an operator which flips
        the third qubit if the first two qubits are `00` or `11`, and does nothing otherwise.
        This can be constructed as
            >>> xor_control_values = cirq.SumOfProducts(((0, 0), (1, 1)))
            >>> q0, q1, q2 = cirq.LineQubit.range(3)
            >>> cirq.X(q2).controlled_by(q0, q1, control_values=xor_control_values)

    2) `(|00><00| + |01><01| + |10><10|) X + (|11><11|) I` represents an operators which flips
        the third qubit if the `nand` of first two qubits is `1` (i.e. first two qubits are
        either `00`, `01` or `10`), and does nothing otherwise. This can be constructed as:
            >>> nand_control_values = cirq.SumOfProducts(((0, 0), (0, 1), (1, 0)))
            >>> q0, q1, q2 = cirq.LineQubit.range(3)
            >>> cirq.X(q2).controlled_by(q0, q1, control_values=nand_control_values)
    """
```

NoureldinYosri replied:
done

daxfohl commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,80 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+        nand =  SumOfProducts(((0, 0), (0, 1), (1, 0))) # nand of two bits
+    """
+
+    _internal_representation: Tuple[Tuple[int, ...], ...]
```
Open question, would this be better as FrozenSet in the outer part, to prevent dupes? It would be slightly slower to iterate and slightly larger to store, so there's a good argument not to change it too. If we do, ProductOfSums could be a Frozenset on the inner part too.

daxfohl reviewed (approved):
Looks good, left a couple questions.

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,80 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+        nand =  SumOfProducts(((0, 0), (0, 1), (1, 0))) # nand of two bits
+    """
+
+    _internal_representation: Tuple[Tuple[int, ...], ...]
```
added, I also test for the uniqueness of products in it

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,80 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+        return frozenset(self._internal_representation) == {(1,) * self._number_variables()}
+
+    def diagram_repr(self) -> str:
+        return NotImplemented
```
the most basic implementation is problematic since the strings will be quite large since they have size num_products x num_qubits => O(num_qubits x 2^num_qubits), this is why I'm hestiant to implement it, however I added it anyway :)

NoureldinYosri commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,80 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+    def __getitem__(
+        self, key: Union[int, slice]
+    ) -> Union['AbstractControlValues', Tuple[int, ...]]:
+        return NotImplemented
```
the reason that I'm hesitant to implement it is that it will have a different meaning from that of ProductOfSums, where in ProductOfSums we get valide values of qubit, here we get a product

tanujkhattar commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,112 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+        This can be constructed as
+        >>> xor_control_values = cirq.SumOfProducts(((0, 0), (1, 1)))
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> cirq.X(q2).controlled_by(q0, q1, control_values=xor_control_values)
```
```suggestion
        >>> xor_cop = cirq.X(q2).controlled_by(q0, q1, control_values=xor_control_values)
```

tanujkhattar commented on cirq-core/cirq/ops/control_values.py:
```diff
@@ -180,3 +180,112 @@ def __and__(self, other: AbstractControlValues) -> 'ProductOfSums':
+
+        >>> nand_control_values = cirq.SumOfProducts(((0, 0), (0, 1), (1, 0)))
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> cirq.X(q2).controlled_by(q0, q1, control_values=nand_control_values)
```
```suggestion
        >>> nand_cop = cirq.X(q2).controlled_by(q0, q1, control_values=nand_control_values)
```

tanujkhattar reviewed (commented):
The doctests are failing because they run the python expression in docstrings written within ">>>" code blocks and compare the expected ouptut with received output. 

I've suggested changes that should fix the failing tests. See https://docs.python.org/3/library/doctest.html for more details.

NoureldinYosri commented:
@tanujkhattar I fixed the docstring and added the optional argument to diagram_repr that we discussed offline

NoureldinYosri commented:
@tanujkhattar  I added the changes we discussed offline namely `and` now works between ProductOfSums and SumOfProducts, and I removed the indexing requirement from the abstrat class and from SumOfProducts, fixed the issues that appeared 

I did't get to testing ControlledOperations/Gates with SumOfProducts, however it should work fine, however I know of a line that will break (and worse silently) with SumOfProducts and that is this one https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/ops/controlled_operation.py#L170 the comparison logic will break I tried a quick fix but it didn't work, I think in a follow up in which we add the tests we discussed we will also need to fix this line
